WordPress/Classes/ViewRelated/Blog/PublicizeConnectionURLMatcher.swift

```diff
@@ -0,0 +1,159 @@
+import Foundation
+
+/// Used to detect whether a URL matches a particular Publicize authorization success or failure route.
+struct PublicizeConnectionURLMatcher {
+    enum MatchComponent {
+        case verifyActionItem
+        case denyActionItem
+        case requestActionItem
+        case stateItem
+        case codeItem
+        case errorItem
+
+        case authorizationPrefix
+        case declinePath
+        case accessDenied
+
+        // Special handling for the inconsistent way that services respond to a user's choice to decline
+        // oauth authorization.
+        // Right now we have no clear way to know if Tumblr fails.  This is something we should try
+        // fixing moving forward.
+        // Path does not set the action param or call the callback. It forwards to its own URL ending in /decline.
+        case userRefused
+
+        // In most cases, we attempt to find a matching URL by checking for a specific URL component
+        fileprivate var queryItem: URLQueryItem? {
+            switch self {
+            case .verifyActionItem:
+                return URLQueryItem(name: "action", value: "verify")
+            case .denyActionItem:
+                return URLQueryItem(name: "action", value: "deny")
+            case .requestActionItem:
+                return URLQueryItem(name: "action", value: "request")
+            case .accessDenied:
+                return URLQueryItem(name: "error", value: "access_denied")
+            case .stateItem:
+                return URLQueryItem(name: "state", value: nil)
+            case .codeItem:
+                return URLQueryItem(name: "code", value: nil)
+            case .errorItem:
+                return URLQueryItem(name: "error", value: nil)
+            case .userRefused:
+                return URLQueryItem(name: "oauth_problem", value: "user_refused")
+            default:
+                return nil
+            }
+        }
+
+        // In a handful of cases, we're just looking for a substring or prefix in the URL
+        fileprivate var matchString: String? {
+            switch self {
+            case .declinePath:
+                return "/decline"
+            case .authorizationPrefix:
+                return "https://public-api.wordpress.com/connect/"
```

Comment:
> Do you know why is the base URL part of the "authorization" check but not part of the "decline" check?

Seems like it's because /decline originates from another host – Path.com. As mentioned in above comment, I think I can look at removing this now.

> Similarly, I'm not sure if it's relevant that "authorization" URL includes a trailing slash.

Yes, I think this might be good to remove!